src/protocol/transport_service.rs

```diff
@@ -122,6 +122,9 @@ pub struct TransportService {
     /// Next substream ID.
     next_substream_id: Arc<AtomicUsize>,
 
+    /// Close the connection if no substreams are open within this time frame.
+    keep_alive: Duration,
```

Comment:
nit: You probably tried to separate this from `keep_alive_timeouts`, but IMO it would be better to use the `_timeout` suffix here as well.